[oeggy03] iP

oeggy03 opened the pull request:
## **EGGBOT** 

> “It's OK to have your eggs in one basket as long as you control what happens to that basket.” – Elon Musk ([source](https://www.brainyquote.com/quotes/elon_musk_567295?src=t_eggs))

EGGBOT frees your mind of having to remember things you need to do. It's,


- text-based
- easy to learn
- SUPER FAST to use

**All you need to do is:**

1. download the jar file from [here](https://github.com/oeggy03/ip/releases/tag/A-Jar).
2. double-click it.
3. type `help` to see all the commands you can use.
4. add your tasks.
5. let it manage your tasks for you 😉

**Or if you want to run it from IntelliJ using gradle...**
1. fork and clone this [repo](https://github.com/oeggy03/ip/).
2. open the folder in IntelliJ.
3. click CTRL twice, then type `gradlew clean run` to run the bot.
4. type `help` to see all the commands you can use.
5. add your tasks.
6. let it manage your tasks for you 😉

Features:

- [ ]  Managing tasks
- [ ]  Managing deadlines 
- [ ]  Reminders (coming soon)

---
For the Java Programmers out their, here's the `main` method!
```
public static void main(String[] args) {
        Duke.initializeBot();

        String introduction = " Hello! I'm EGGBOT!\n" +
                " Type 'help' to show available commands!";

        printBasicOutput(introduction);
        Scanner scanner = new Scanner(System.in);

        while (isRunning) {
            System.out.println("Please type a command, or type 'help' to show available commands:");

            String input = scanner.nextLine().strip();  // Read input
            Duke.commandHandler.parseInput(input);
        }

    }
    ```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
Finished Level-0
Renamed the bot
Implement an initial skeletal version of the it that simply greets the user and exits.

Commit:
Level 1 done: Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

Commit:
Level 2 done: Add the ability to store whatever text entered by the user and display them back to the user when requested.

Commit:
Done Level 3: Add the ability to mark tasks as done, add the ability to change the status back to not done.

Commit:
Added A-Classes - Task Class
For Level 3

Commit:
Added classes for Deadline, Event, ToDo
Added error handling for bad user input for deadline/event which might break program
Finished Level 4

Commit:
Done with: Use the input/output redirection technique to semi-automate the testing of Duke.

Commit:
Added a new exception: DukeInvalidFormatException, to handle cases of deadline/event not having proper inputs separated by '/'
Common errors handled
Tests updated to accommodate changes
Level 5: Error Handling done

Commit:
Edited current code to use ArrayList<Task> instead

Commit:
Done with level 6
Delete command works

Commit:
Updated the Automated testing to accommodate changes for level 6

Commit:
Enum is not applicable as infinitely different inputs possible when deleting/marking/unmarking/adding tasks.
I chose to use startswith() instead, easier, makes more sense

Commit:
Added memory.txt, to store taks

Commit:
Added memory.txt as data storage for chatbot
Created TaskArray object for array of tasks

Commit:
Major restructuring of project:
Created Database package and Dbops.java to handle all communication with database.
Grouped all Exceptions into Exceptions package.
Created Printers package and various printer classes to handle all output from the bot to ensure consistency in output.
Grouped all Task Models and TaskArray into Models package.
Improved exception handling for absence/corruption of memory.txt, the bot's memory file

Commit:
Revamped Code Structure: Added Handlers Package, and separate handlers for each user command, including a handler for invalid input.
Added a Central Command Handler to handle all user inputs, and then assign individual handlers to them.

Commit:
Added a "help" command to show a list of commands.
Fixed minor output bugs.

Commit:
Fixed bug where first task will delete itself if program terminates unexpectedly.
Added DateTime support, where Event and Deadline tasks take in dd-MM-yyyy Hmm format.

Commit:
Edited the help command to also display information on the datetime format.

Commit:
Merge branch 'branch-level-8'

Commit:
Updated Javadocs for DateTime Parser

Commit:
OOP has already been done prior to this increment.
Renamed Printer package to Ui, and Database package to Storage

Commit:
Merge branch 'branch-a-moreoop'

Commit:
All classes are already organised in appropriate packages.

Commit:
Merge branch 'master' into add-gradle-support

brein62 commented on src/main/java/LogicHandlers/CommandHandlers/Command.java:
```diff
@@ -0,0 +1,11 @@
+     *
+     * @param commandContent The content of the input.
+     */
+    void parseCommandContent(String commandContent);
```
Probably unintended but I think you intend for this to be `public`? It's mostly for code consistency (interface methods are by default `public abstract`)

brein62 commented on src/main/java/LogicHandlers/CommandHandlers/EventHandler.java:
```diff
@@ -0,0 +1,64 @@
+                }
+
+                tasks.addTask(new Event(taskArr[0].strip(), false,
+                        parseDateTimeFromString(taskArr[1].strip()), parseDateTimeFromString(taskArr[2].strip())));
```
woah this code is quite trippy to read. Probably can split into different expressions to ease readability:

```java
String eventDescription = taskArr[0].strip();
LocalDateTime fromDate = parseDateTimeFromString(taskArr[1].strip());
LocalDateTime toDate = parseDateTimeFromString(taskArr[2].strip());

tasks.addTask(new Event(eventDescription, false, fromDate, toDate));
```

brein62 commented on src/main/java/LogicHandlers/CommandHandlers/ExitHandler.java:
```diff
@@ -0,0 +1,35 @@
+    public void parseCommandContent(String commandContent) {
+        printBasicOutput("Bye. Hope to see you again soon!");
+
+        exit(0);
```
Interesting choice to quit the app directly from the `ExitHandler` :thinking:

brein62 commented on src/main/java/Models/TaskArray.java:
```diff
@@ -0,0 +1,89 @@
+
+    public void modifyTask(int index, Task modifiedTask) {
+
+    }
```
Empty function body ❓ 
Can be removed if you're not planning to implement this method in the future

brein62 commented on src/main/java/Storage/Dbops.java:
```diff
@@ -0,0 +1,128 @@
+     * @param taskArray The TaskArray to load the tasks from.
+     */
+    public static void saveTasksToFile(TaskArray taskArray) {
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(Dbops.FILEPATH + Dbops.FILENAME))) {
```
Wrong `try` implementation syntax, making the `try` statement look like a `while` loop lmao

You probably meant:
```java
try {
    BufferedWriter writer = new BufferedWriter(new FileWriter(Dbops.FILEPATH + Dbops.FILENAME));
    
    for (Task task : taskArray) {
        // code
    }
} catch (IOException e) {
    // code
}
```

brein62 reviewed (commented):
Code and overall implementation LGTM. Code quality overall is really good and I really like the heavy use of OOP in your implementation. Good job! 👍

Commit:
Added build support for iP project with Gradle

Commit:
Merge branch 'add-gradle-support'

Choonyan02 commented on src/main/java/Models/Task.java:
```diff
@@ -0,0 +1,31 @@
+
+public class Task {
+    Boolean isMarked;
+    String name;
```
Perhaps you can indicate the fields as protected fields?

Choonyan02 commented on src/main/java/Models/Deadline.java:
```diff
@@ -0,0 +1,30 @@
+import static LogicHandlers.Parsers.DateTimeParser.parseStringFromDateTime;
+
+public class Deadline extends Task {
+    LocalDateTime endTime;
```
Perhaps you can encapsulate the class field here by making them private?

Choonyan02 commented on src/main/java/Models/Event.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Event extends Task {
+    LocalDateTime startTime;
+    LocalDateTime endTime;
```
Same as Deadline, perhaps you can encapsulate the class fields here by making them private?

Choonyan02 commented on src/main/java/Exceptions/DukeCorruptedDataException.java:
```diff
@@ -0,0 +1,8 @@
+package Exceptions;
```
I like that you have a lot of exceptions, which makes it clearer to user what went wrong 😄

Choonyan02 commented on src/main/java/LogicHandlers/CommandHandlers/DeadlineHandler.java:
```diff
@@ -0,0 +1,63 @@
+                tasks.addTask(new Deadline(taskArr[0].strip(),
+                        false, parseDateTimeFromString(taskArr[1].strip())));
+
+                String output = "Got it, I've added this task: \n" +
+                        tasks.get(tasks.size() - 1) + "\n" +
+                        "You now have " + tasks.size() + " tasks in the list";
+
+                printBasicOutput(output);
```
Perhaps you can leave it to Ui to do this?

Choonyan02 commented on src/main/java/LogicHandlers/CommandHandlers/DeadlineHandler.java:
```diff
@@ -0,0 +1,63 @@
+                printBasicOutput(output);
+
+            } catch (DukeInvalidFormatException | DukeInvalidDateTimeException e) {
+                String errorString = "Something went wrong! Please format the task properly and add it again. \n" +
+                        "Error: " + e;
+
+                printErrorOutput(errorString);
```
Same with this, perhaps you can leave it to the Ui to print?

Choonyan02 reviewed (commented):
Overall, LGTM! Very good OOP and the directories are neat too.

Commit:
Restructured CommandHandlers to return outputs, which will then be printed by OuputePrinters, instead of CommandHandlers printing it themselves.
Created JUnit Test for parseCommandContent for TodoHandler.

Commit:
Added JUnit Tests for DeadlineHandler

Commit:
Merge branch 'add_junit_tests'

Commit:
Moved the data folder out of src.
Added a mkdir command in Dbops, to handle cases where

Commit:
Added support for find, and handling of tasks not found for find

Commit:
Added support for JavaFX
Built a responsive UI for Eggbot
Renamed Duke.java to Eggbot.java to properly reflect the name of the bot
Revised the printer classes into one printer class for better abstraction

Commit:
Merge branch 'branch-level-10'

Commit:
Update build.gradle

Set mainClass to Launcher to adapt for JavaFX updates

Commit:
Modify build.gradle

Previously, assertions were not enabled and this prevented the use of assertions, therefore critical errors and input requirements could not be swiftly caught.

Now, assertions have been enabled to allow for the use of the assert function, to catch these errors quickly.

Commit:
Add assertions to confirm that inputs for Todo, Deadline or Event tasks are not null

Previously, if else statements were used to determine whether the input was valid (i.e. not null). This caused more nested code and reduced the code quality, and prevented us from checking for input validity efficiently.

Add assert statements to return an assertion error when the user inputs no arguments for a Todo, Deadline or Event task.

This allows for inputs to be checked in one line, and reduces nested code.

Commit:
Update tests for DeadlineHandler and TodoHandler

Tests have been updated to match the outputs of the latest update of DeadlineHandler and TodoHandler, and now recgonises the assertion error.

Commit:
Update access modifiers for Task, Deadline, Event

Previously, there were no access modifiers for some variables in the mentioned classes. This may lead to improper usage of private class variables.

Update the access modifiers for certain variables to be protected or private, as appropriate. Prevents unintentional breaking of the abstraction barrier.

As a step towards prevent unintentional breaking of the abstraction barrier and improper usage of supposedly private variables, let's be more mindful and remember to assign appropriate access variables when creating new variables and methods.

Commit:
Update the try-catch syntax for method saveTasksToFile

Previously, the syntax was implemented in a way which might be confusing to read.

The syntax has been changed to fit that taught in CS2103T.

oeggy03 committed:
Merge pull request #2 from oeggy03/branch-assertions

Added assertions to code

oeggy03 committed:
Merge pull request #3 from oeggy03/branch-code-quality

Improved code quality for various methods and classes

Commit:
Update help command

Help command string is now marked as private

Commit:
Merge branch 'branch-help-extension'

Commit:
Update help command to include numbering

Numbering of commands helps the user to read the commands more easily.

Commit:
Merge branch 'branch-help-extension'

Commit:
Update egg.png

Updated the profile picture of EGGBOT to make it nicer.

Commit:
Merge branch 'branch-better-gui' into branch-code-quality

Commit:
Update package names

Previously, packages did not follow the naming conventions of the coding standard due to them starting with capital letters.

They now are all in lowercase.

Commit:
Merge branch 'branch-code-quality'

Commit:
Add Ui.png

Added Ui.png as a representative screenshot.

Commit:
Update DeadlineHandlerTest

Updated it to match the new package names.

Commit:
Update memory.txt

Commit:
Update Dbops to fix bug where tasks were not saving

There used to be a bug where tasks added by the user are not saved to memory.txt, due to me implementing a change suggested by one of my PR commenters.

The bug has now been solved by me reverting to the original statement.

Commit:
Merge branch 'branch-help-extension' into branch-code-quality

# Conflicts:
#	data/memory.txt

Commit:
Remove memory.txt after debugging

Commit:
Merge branch 'branch-code-quality'

Commit:
Fix grammatical error in Eggbot's response when adding tasks.

Previously, Eggbot would reply with "1 tasks" when listing total number of tasks.

Now, Eggbot will use "task(s)" instead to avoid the error.

Commit:
Update docs/README.md to include user guide

Updated the README.md to include the user guide, in order to help users familiarise themselves with the bot's commands.

Commit:
Add a sample memory.txt file

Previously, new users would not have any existing tasks to try out the commands with.

Now, the sample memory.txt will provide 3 starting tasks (1 of each type) for new users to experiment with.

Commit:
Update DeadlineHandlerTest and TodoHandlerTest

Updated the tests to match the new outputs given by the reformatted reply message (which uses task(s) instead of tasks)

Commit:
Add lines between commands to improve readability

Before, it was difficult to discern between the different sections for different commands in the Features and Commands section.

Now, the line should help readers see the different sections for the different commands better.

Commit:
Update README.md to include Ui.png

Allows for user to get a view of the UI of Eggbot when they access the user guide.

Commit:
Merge branch 'master' into branch-better-gui

Commit:
Update the design of eggbot majorly

Eggbot now has a new colour scheme and better button and speech bubble designs.

Commit:
Delete memory.txt after testing

Commit:
Merge branch 'branch-better-gui'

Commit:
Merge branch 'master' into branch-user-guide

Commit:
Update README.md to include updated Eggbot UI

Commit:
Merge branch 'branch-user-guide'

Commit:
Update to increase width of Ui.png in README.md

Commit:
Merge branch 'master' into branch-better-gui

Commit:
Add Easter Egg section and improve getting started section

Commit:
Merge branch 'branch-better-gui'

Commit:
Add crack.png to images

Commit:
Delete Easter Egg from Readme.md

Commit:
Update build version to v0.3

Commit:
Update README.md in main directory

This prevents the confusion that the bot is still named Duke.
